# 01 — Data Exploration

This notebook covers all Day 1 exploration work:
1. Load and profile the Vickers & Vertosick (2016) dataset
2. EDA — distributions, relationships, and visualizations
3. Profile the Afonseca et al. training dataset

**Primary dataset:** Vickers & Vertosick — 2,303 recreational runners with self-reported training and race times.  
**Secondary dataset:** Afonseca et al. — 10.7M training records from 36,412 World Marathon Majors runners.

## Setup

In [4]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os

from src.data_loading import load_vickers

sns.set_theme(style='whitegrid')
%matplotlib inline

---
## Part 1: Vickers Dataset — Load and Profile

Load the XLSX, standardize column names to snake_case, and check structure.

In [9]:
df = load_vickers()
print(f'Shape: {df.shape}')
df.dtypes.to_frame('dtype')

Shape: (2303, 50)


dtype
id                int64
adjusted          int64
age             float64
bmi             float64
cohort1         float64
cohort2         float64
cohort3         float64
cohort4         float64
endurancecat      int64
endurancespeed    int64
female            int64
footwear          int64
group           float64
injury            int64
k5_d            float64
k5_di           float64
k5_ti           float64
k5_ti_adj       float64
k5_tr           float64
k10_d           float64
k10_di          float64
k10_ti          float64
k10_ti_adj      float64
k10_tr          float64
m5_d            float64
m5_di           float64
m5_ti           float64
m5_ti_adj       float64
m5_tr           float64
m10_d           float64
m10_di          float64
m10_ti          float64
m10_ti_adj      float64
m10_tr          float64
mh_d            float64
mh_di           float64
mh_ti           float64
mh_ti_adj       float64
mh_tr           float64
mf_d            float64
mf_di           float64
mf_ti           float64
mf_ti_adj       float64
mf_tr           float64
max             float64
sprint            int64
tempo             int64
typical         float64
model1_time     float64
model2_time     float64

Ok, all int64 and float64. I know that some of these columns are still categorical based on the docs though. 

In [10]:
df.describe().T

count          mean          std           min           25%  \
id              2303.0   1154.437256   667.163997      1.000000    577.500000   
adjusted        2303.0      0.939644     0.238197      0.000000      1.000000   
age             2303.0     36.421407     9.913292     16.000000     29.000000   
bmi             2303.0     23.745838     3.052466     15.428386     21.704651   
cohort1          929.0      1.000000     0.000000      1.000000      1.000000   
cohort2          633.0      1.000000     0.000000      1.000000      1.000000   
cohort3          493.0      1.000000     0.000000      1.000000      1.000000   
cohort4          387.0      1.000000     0.000000      1.000000      1.000000   
endurancecat    2303.0      1.833695     0.656222      1.000000      1.000000   
endurancespeed  2303.0      3.440729     1.768493      1.000000      2.000000   
female          2303.0      0.386452     0.487042      0.000000      0.000000   
footwear        2303.0      1.812419     0.425596      1.000000      2.000000   
group           2164.0      2.000462     0.816213      1.000000      1.000000   
injury          2303.0      1.410769     0.649605      1.000000      1.000000   
k5_d            1387.0   5000.000000     0.000000   5000.000000   5000.000000   
k5_di           1387.0      3.146359     0.800112      1.000000      3.000000   
k5_ti           1387.0   1402.029200   325.815925    852.000000   1168.500000   
k5_ti_adj       1289.0   1398.790166   321.199510    855.466614   1166.435547   
k5_tr           1387.0      2.018745     0.757603      1.000000      2.000000   
k10_d            946.0  10000.000000     0.000000  10000.000000  10000.000000   
k10_di           946.0      2.950317     0.862300      1.000000      2.000000   
k10_ti           946.0   3000.086681   680.762376   1602.000000   2519.250000   
k10_ti_adj       883.0   3003.035021   670.427533   1789.104614   2520.968018   
k10_tr           946.0      2.006342     0.734603      1.000000      2.000000   
m5_d             313.0   8045.000000     0.000000   8045.000000   8045.000000   
m5_di            313.0      2.900958     0.831915      1.000000      2.000000   
m5_ti            313.0   2347.322684   615.359006   1277.000000   1910.000000   
m5_ti_adj        295.0   2353.981723   605.359251   1309.209961   1913.209534   
m5_tr            313.0      2.092652     0.738834      1.000000      2.000000   
m10_d            357.0  16090.000000     0.000000  16090.000000  16090.000000   
m10_di           357.0      2.913165     0.986294      1.000000      2.000000   
m10_ti           357.0   4926.886555  1095.178674   2875.000000   4142.000000   
m10_ti_adj       313.0   4945.558649  1049.846384   3105.000000   4200.000000   
m10_tr           357.0      1.896359     0.672838      1.000000      1.000000   
mh_d            1579.0  21097.500000     0.000000  21097.500000  21097.500000   
mh_di           1579.0      2.907536     0.868583      1.000000      2.000000   
mh_ti           1579.0   6568.374921  1343.401561   3580.000000   5603.500000   
mh_ti_adj       1444.0   6582.467434  1338.185551   3640.434814   5615.750000   
mh_tr           1579.0      1.853705     0.681647      1.000000      1.000000   
mf_d            1022.0  42195.000000     0.000000  42195.000000  42195.000000   
mf_di           1022.0      2.862035     0.892644      1.000000      2.000000   
mf_ti           1022.0  13507.810176  2705.690550   8152.000000  11580.000000   
mf_ti_adj        929.0  13448.592130  2628.031295   8217.089844  11580.000000   
mf_tr           1022.0      1.684932     0.644838      1.000000      1.000000   
max             2303.0     40.696709    21.515680      0.000000     25.000000   
sprint          2303.0      0.512809     0.499944      0.000000      0.000000   
tempo           2303.0      0.576639     0.494199      0.000000      0.000000   
typical         2303.0     32.530195    17.142133      0.000000     20.000000   
model1_time      310.0    222.730997    39.23257

Mean mileage per week was 32 miles, but standard deviation was 17 miles. This means we had significant variaition in mileage ran per week. I should look over that more closely when I do thorough EDA.

In [15]:
df.isnull().sum().to_frame("null count")

null count
id                       0
adjusted                 0
age                      0
bmi                      0
cohort1               1374
cohort2               1670
cohort3               1810
cohort4               1916
endurancecat             0
endurancespeed           0
female                   0
footwear                 0
group                  139
injury                   0
k5_d                   916
k5_di                  916
k5_ti                  916
k5_ti_adj             1014
k5_tr                  916
k10_d                 1357
k10_di                1357
k10_ti                1357
k10_ti_adj            1420
k10_tr                1357
m5_d                  1990
m5_di                 1990
m5_ti                 1990
m5_ti_adj             2008
m5_tr                 1990
m10_d                 1946
m10_di                1946
m10_ti                1946
m10_ti_adj            1990
m10_tr                1946
mh_d                   724
mh_di                  724
mh_ti                  724
mh_ti_adj              859
mh_tr                  724
mf_d                  1281
mf_di                 1281
mf_ti                 1281
mf_ti_adj             1374
mf_tr                 1281
max                      0
sprint                   0
tempo                    0
typical                  0
model1_time           1993
model2_time           2132

Interesting. The most submitted race was the half-marathon followed by the 5k, with the least submitted races being the 5 mile and the 10 mile. Approximately 40% of runners submitted a marathon time, which is promising. Lets see how many of these marathon runners have other race times logged. 

In [20]:
has_marathon = df[df["mf_ti_adj"].notna()]

race_cols = ["k5_ti_adj", "k10_ti_adj", "m5_ti_adj", "m10_ti_adj", "mh_ti_adj"]
has_marathon["other_races"] = has_marathon[race_cols].notna().sum(axis=1)

has_marathon["other_races"].value_counts().sort_index().to_frame("# of other races")

# of other races
other_races                  
1                         436
2                         493

Alright, so we have 493 runners who have an adjusted marathon time, who also ran two other races with adjusted times. That shaves n down significantly.

### Data quality checks

I'm looking for dupes and negatives, along with impossible values. All of this was self reported to a magazine, so I want to be diligent.

In [21]:
# Duplicates
print(f'Duplicate rows: {df.duplicated().sum()}')

Duplicate rows: 0


In [25]:

impossible = df[(df["mf_ti_adj"] < 120*60) | (df["mf_ti_adj"] > 480*60)]
print(f"Total impossible vals: {len(impossible)}")


Total impossible vals: 0


Awesome. No impossible values. Vickers et al. Almost certainly cleaned this data before they used it, but its always good to check.

### Vickers profiling summary

We have 493 datapoints to train off of, which should be a decent amount for our purposes. Our data seems pretty clean already. We have no dupe rows and no impossible marathon values in our data. Based on this, its worth exploring the data further. I want to inspect the distribution of training volume, since thats a major feature. Its also worth inspecting what values are null in the data we will be working with to make sure theres no systemic issues. EDA is called for next.

---
## Part 2: EDA — Distributions and Relationships

### Marathon finish time distribution

In [ ]:
# Histogram of marathon finish times
fig, ax = plt.subplots(figsize=(10, 5))
# sns.histplot(df['marathon_time'], kde=True, ax=ax)
# ax.set_xlabel('Marathon Finish Time (min)')
# ax.set_title('Distribution of Marathon Finish Times')
# plt.tight_layout()
# plt.savefig('../results/figures/marathon_time_dist.png', dpi=150)

*What do you see? Is it normally distributed? Skewed?*

### Weekly mileage distribution

In [ ]:
# Histogram of weekly mileage
fig, ax = plt.subplots(figsize=(10, 5))
# sns.histplot(df['weekly_mileage'], kde=True, ax=ax)
# ax.set_xlabel('Weekly Mileage')
# ax.set_title('Distribution of Weekly Training Mileage')
# plt.tight_layout()
# plt.savefig('../results/figures/weekly_mileage_dist.png', dpi=150)

*What do you see?*

### Core relationship: weekly mileage vs marathon time

In [ ]:
# Scatter plot — this should show a clear negative relationship
fig, ax = plt.subplots(figsize=(10, 6))
# sns.scatterplot(data=df, x='weekly_mileage', y='marathon_time', alpha=0.5, ax=ax)
# ax.set_xlabel('Weekly Mileage')
# ax.set_ylabel('Marathon Time (min)')
# ax.set_title('Weekly Mileage vs Marathon Finish Time')
# plt.tight_layout()
# plt.savefig('../results/figures/mileage_vs_marathon.png', dpi=150)

*Describe the relationship. Be specific — e.g., "runners averaging 40+ miles/week finish ~30 min faster than those under 20 mpw."*

### Half marathon time vs marathon time

In [ ]:
# Scatter plot — should show strong positive correlation
fig, ax = plt.subplots(figsize=(10, 6))
# sns.scatterplot(data=df, x='half_marathon_time', y='marathon_time', alpha=0.5, ax=ax)
# ax.set_xlabel('Half Marathon Time (min)')
# ax.set_ylabel('Marathon Time (min)')
# ax.set_title('Half Marathon Time vs Marathon Finish Time')
# plt.tight_layout()
# plt.savefig('../results/figures/half_vs_marathon.png', dpi=150)

*What do you see?*

### Marathon time by sex and age group

In [ ]:
# Box plot — marathon time by sex
fig, axes = plt.subplots(1, 2, figsize=(14, 5))
# sns.boxplot(data=df, x='sex', y='marathon_time', ax=axes[0])
# axes[0].set_title('Marathon Time by Sex')

# Box plot — marathon time by age group (you may need to bin ages first)
# sns.boxplot(data=df, x='age_group', y='marathon_time', ax=axes[1])
# axes[1].set_title('Marathon Time by Age Group')
# plt.tight_layout()
# plt.savefig('../results/figures/marathon_by_demographics.png', dpi=150)

*What do you see?*

### Correlation heatmap

In [ ]:
# Correlation heatmap of all numeric features
fig, ax = plt.subplots(figsize=(12, 8))
# corr = df.select_dtypes(include=np.number).corr()
# sns.heatmap(corr, annot=True, fmt='.2f', cmap='coolwarm', center=0, ax=ax)
# ax.set_title('Feature Correlation Heatmap')
# plt.tight_layout()
# plt.savefig('../results/figures/correlation_heatmap.png', dpi=150)

*Which features are most correlated with marathon time? Which features are highly correlated with each other (redundant)?*

### Scatter matrix of top features

In [ ]:
# Pairplot of the 4-5 most promising features
# top_features = ['marathon_time', 'half_marathon_time', 'weekly_mileage', '10k_time', 'bmi']
# sns.pairplot(df[top_features].dropna(), corner=True)
# plt.savefig('../results/figures/pairplot_top_features.png', dpi=150)

*What do you see?*

### EDA summary

*Write a paragraph summarizing the key relationships and takeaways from the visualizations above.*

---
## Part 3: Afonseca Training Dataset — Profile

Quick look at the large-scale training data to decide whether it's usable within the 5-day timeline.

In [ ]:
# Load the weekly-level Parquet file
# af = pd.read_parquet('../data/raw/afonseca/weekly.parquet')
# print(f'Shape: {af.shape}')
# af.dtypes

In [ ]:
# af.describe()

In [ ]:
# How many unique athletes? How many have marathon major participation?
# print(f'Unique athletes: {af["athlete"].nunique()}')

In [ ]:
# Distribution of weekly training distance and duration

### Key limitation

The Afonseca dataset records which marathon(s) each athlete participated in but **does not include actual finish times**. To use this data for prediction, you would need to cross-reference with scraped race results — non-trivial given athletes are anonymized.

### Decision: use this dataset or set it aside?

*Write your decision and reasoning here. Update `data/README.md` accordingly.*